jep/0000/README.adoc

```diff
@@ -0,0 +1,365 @@
+Since its origins in Hudson, this was using a fork of the upstream library maintained by the Jenkins project.
+That fork is now deprecated, and Jenkins core integrates the latest official release instead.
+
+== Specification
+
+`org.jvnet.hudson:xstream:1.4.7-jenkins-1` is now replaced by the latest version of `com.thoughtworks.xstream:xstream`, as of this writing `1.4.13`.
+
+Automatic dynamic detection of XStream annotations was problematic and has been removed.
+
+If JEP-7 is implemented, then `XStream2.setMapper` and `.getMapperInjectionPoint` can be removed as `ruby-runtime` was the only caller.
+
+== Motivation
+
+Maintaining a custom fork of a library was an additional maintenance burden for Jenkins developers.
+The fork was well out of date, leading to suspicions that security hardenings added in recent years may be missing, though no vulnerabilities are known.
+
+A key reason for maintaining the fork was to introduce thread safety as well as concurrency-optimized collections into XStream.
+In fact upstream XStream is designed to be thread-safe _after_ its configuration has been frozen.
+However, Jenkins was relying on autodetection of XStream annotations during serialization (and, with some limitations, deserialization).
+This forced the fork to be patched to prevent race conditions during serialization when annotations were newly discovered.
+The XStream maintainer never agreed with this design and refused to integrate these patches.
+This complicated maintenance of the fork, and may have led to race conditions and/or blocked threads at runtime.
+
+== Reasoning
+
+The original plan for dealing with thread safety was to keep a pool of `XStream2` instances backed by each public field of this type.
+The pooled instances, each limited to one thread at a time, would clone configuration from the main instance.
+This however seemed rather complex (there is no simple interface to implement as a proxy—`XStream` is a class with plenty of state),
+and performance may have suffered from keeping these pools and duplicating various caches.
+Since the number of plugins actually using XStream annotations was fairly small,
+and adapting to the change can be done in as little as one line of code,
+it seems simpler and better to use XStream the way its maintainer intended.
+
+`ToolInstaller.properties` was tough to deal with.
+This field must be serialized via XStream (for example it lists `ToolInstallation` instances),
+but cannot be serialized over Remoting
+(neither its collection type, nor typical element types, are or could reasonably be made `Serializable`).
+Nor did it seems reasonable to mark the field `transient` but define a custom XStream converter, since subclasses have their own fields.
+Fortunately not a lot of plugins seem to attempt to send a `ToolInstaller` over Remoting;
+all the known cases involve some agent-side computation that actually only needs the tool home directory.
```

Comment:
It may worth double-checking Custom Tools Plugin. I do not recall logic there, but has quite complex tool initialization logic

Reply:
Handled by https://github.com/jenkinsci/custom-tools-plugin/pull/51 I believe.